### PR TITLE
Refactor board from String[] to ArrayList<String> 

### PR DESCRIPTION
**Pull request by martingaston**
This is a pretty big PR because, as it (expectedly) turns out - handing out a `String[]` meant that app was very brittle and resistant to change. The rumours were true! 

The single 'conversion' point is now player.getSymbol() which returns a string of the player's symbol (e.g. X or O). 

This introduces a new class, `Cell`, which is largely what gets passed around between classes. 

One additional way of trying to insulate from app-breaking changes is that I've tried to put the bits where the classes get convert that input into their own private methods, so if `getOccupant` 

```
private String renderCell(Cell cell) {
        return cell.getOccupant();
    }
```

The logic here being that if getOccupant were to change (I'm still a bit unsure about whether to return the string or the `Player` class, for instance) then it doesn't go straight into the affected methods. 

I do think this opens up some additional problems with regards to Single Responsibility - the `Player` class both looks after the Player's symbol **and** provides the interface for getting input from said player, and with the `Player` class being passed around the application it means many classes have the potential to get input, if they so desired. This is something I will hopefully have time to look at further!

### CODE REVIEW

File: src/main/java/Board.java [hunk 1]
```diff
@@ -1,39 +1,34 @@
+import java.util.ArrayList;
+import java.util.Set;
+
 public class Board {
-    private String[] board;
+    private int totalCells = 9;
+    private ArrayList<Cell> board = new ArrayList<>(totalCells);
```
Comment (felipesere) [inline src/main/java/Board.java]: A good old-fashioned trick to keep some design options open is to have interfaces on the left of the assignment and implementations on the right. The interface you are looking for is List<T>. 

It also means you don’t have to repeat Array... all the time 😅

File: src/main/java/Cell.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
+public class Cell {
+    private Player occupant;
+
+    String getOccupant() {
+        if (isNotOccupied()) {
+            return " ";
```
Comment (felipesere) [inline src/main/java/Cell.java]: Is this empty string critical for printing/formatting?

One _option_ is to use an Optional<String> (♥️) to express the idea that “there may or may not be a string here, please check!” and letting the caller decide what to do when there isn’t one. 

This alleviates a very subtle way of coupling due to defaults. It makes sense for the visual aspect of TTT to default unoccupied cells to “ “. But for the negamax part it means you need to infer “when it is ‘’ then it means it’s empty and I can place my mark there”. Or when you change to a 4x4 you’ll need to upgrade this class to default to “  “.

File: src/main/java/Display.java [hunk 3]
```diff
@@ -15,53 +17,55 @@ public Display(Board board) {
...
 
-        return String
-                .format("| %s | %s | %s |", left, center, right)
-                .concat(newline)
-                .concat(divider);
+    private String renderCell(Cell cell) {
+        return cell.getOccupant();
     }
+
+    private int zeroIndexToOneIndex(int zeroIndexedNumber) { return zeroIndexedNumber + 1; }
```
Comment (felipesere) [inline src/main/java/Display.java]: Happy to see you pulled out something as harmless-looking as +1 into its own function.   In my option the world needs more names to allow the reader to understand the intent of things. ♥️ 

I suggest we look for a name that doesn’t “just” replicate its implementation though. We do we add 1? It’s for us humans! We don’t really count from 0 upwards, so this function there to “humanize(int position)” 😍

File: src/main/java/Board.java (hunk 1 above)
Comment (martingaston) [reply src/main/java/Board.java]: Interesting... Interfaces are still new to me (it's taken me two weeks to realise why IntelliJ colours some types as green 🤦‍♂️ ) but this makes a lot of sense!

File: src/main/java/Display.java (hunk 3 above)
Comment (martingaston) [reply src/main/java/Display.java]: That makes good sense, and then I suppose you can keep the variable more specific as well so it it helps self-document!

File: src/main/java/Cell.java (hunk 2 above)
Comment (martingaston) [reply src/main/java/Cell.java]: I'm looking at the API docs for this now and I like it, I am going to work on this as a separate PR 👍

**Pull request merged**